edc-extensions/cx-policy/src/main/java/org/eclipse/tractusx/edc/policy/cx/common/AbstractVpConstraintFunction.java

```diff
@@ -0,0 +1,135 @@
+     * @param credentialType the credential type that will be verified against.
+     */
+    public AbstractVpConstraintFunction(String credentialType) {
+        requireNonNull(credentialType);
+        this.credentialType = credentialType;
+        this.errorPrefix = format(ERROR_PREFIX_TEMPLATE, credentialType);
+    }
+
+    /**
+     * Validates the operator is in the set of expected operators.
+     */
+    protected boolean validateOperator(Operator operator, PolicyContext context, Operator... expectedOperators) {
+        var set = stream(expectedOperators).collect(Collectors.toSet());
+        if (!set.contains(operator)) {
+            var valid = set.stream().map(Enum::toString).collect(joining(","));
+            context.reportProblem(format("Unsupported operator for %s credential constraint, only %s allowed: %s", credentialType, valid, operator));
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Validates the VP by checking that it is a {@link JsonObject}.
+     */
+    protected boolean validatePresentation(@Nullable Object vp, PolicyContext context) {
+        if (vp == null) {
+            context.reportProblem(format("%s VP not found", credentialType));
+            return false;
+        }
+
+        if (!(vp instanceof JsonValue jsonValue)) {
+            context.reportProblem(format("%s VP is not a JSON type: %s", credentialType, vp.getClass().getName()));
+            return false;
+        }
+
+        if (!(OBJECT == jsonValue.getValueType())) {
+            context.reportProblem(format("%s VP must be type %s but was: %s", credentialType, OBJECT, jsonValue.getValueType()));
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Returns the credential subject portion of a VC or null if there was an error. Error information will be reported to the context.
+     */
+    @Nullable
+    protected JsonObject extractCredentialSubject(JsonObject credential, PolicyContext context) {
+        var subjectArray = credential.get(CREDENTIAL_SUBJECT);
+        if (subjectArray == null || subjectArray.getValueType() != ARRAY) {
+            context.reportProblem(errorPrefix + " no credentialSubject found");
+            return null;
+        }
+        if (subjectArray.asJsonArray().size() != 1) {
+            context.reportProblem(errorPrefix + " empty credentialSubject");
+            return null;
+        }
+
+        var subjectValue = subjectArray.asJsonArray().get(0);
+        if (subjectValue == null || subjectValue.getValueType() != OBJECT) {
+            context.reportProblem(errorPrefix + " invalid credentialSubject format");
+            return null;
+        }
+
+        return subjectValue.asJsonObject();
+    }
+
+    /**
+     * Returns true if the actual operand value is a string literal case-insensitive equal to the expected value.
+     */
+    protected boolean validateRightOperand(String expectedValue, Object actualValue, PolicyContext context) {
```

Comment:
I think the expected value should be a vararg, because there could be several allowed ones, same as we have for the `validateOperator`

Reply:
> I think the expected value should be a vararg, because there could be several allowed ones, same as we have for the `validateOperator`

Yeah, that method is going to actually have to be modified when we switch versions from the left to right operand as described [here](https://github.com/eclipse-tractusx/ssi-docu/pull/9). I'll do that in another PR.